### PR TITLE
Data source prefetching support

### PR DESCRIPTION
**Pull request by RowbotNZ**
I have added reactive wrappers for the `prefetchDataSource` delegate on `UICollectionView` and `UITableView` (added in iOS 10 and tvOS 10). 

I have tried to follow as closely as possible the examples set out in `RxCollectionViewDataSourceProxy` and `RxScrollViewDelegateProxy`, while lacking an in-depth understanding of the underlying `DelegateProxy` class. Please let me know if my implementation is misusing the `DelegateProxy` superclass.

I have also written some very basic unit tests for the new functionality that I have introduced, as per the contribution guidelines.

Also as per the contribution guidelines, I tried to run the test script, `./scripts/all-tests.sh`, however it failed with a Swift runtime error. I then attempted to run the test script on `develop` (commit `8a413e8fa860db6f894b0593105f15940bb833db`) and got the same error. If there is anything I can do to fix this please let me know.

### CODE REVIEW

File: RxCocoa/iOS/Proxies/RxCollectionViewDataSourcePrefetchingProxy.swift [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import UIKit
+import RxSwift
+
+@available(iOS 10.0, tvOS 10.0, *)
+extension UICollectionView: HasPrefetchDataSource {
+    public typealias PrefetchDataSource = UICollectionViewDataSourcePrefetching
+}
+
+@available(iOS 10.0, tvOS 10.0, *)
+let collectionViewPrefetchDataSourceNotSet = CollectionViewPrefetchDataSourceNotSet()
```
Comment (kzaher) [inline RxCocoa/iOS/Proxies/RxCollectionViewDataSourcePrefetchingProxy.swift]: I think that we can make this `fileprivate`. (and also make the other corresponding fields like `collectionViewDataSourceNotSet` `fileprivate`).

File: RxCocoa/iOS/Proxies/RxCollectionViewDataSourcePrefetchingProxy.swift [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+@available(iOS 10.0, tvOS 10.0, *)
+final class CollectionViewPrefetchDataSourceNotSet : NSObject, UICollectionViewDataSourcePrefetching {
+
+    func collectionView(_ collectionView: UICollectionView, prefetchItemsAt indexPaths: [IndexPath]) {}
+
+}
+
+@available(iOS 10.0, tvOS 10.0, *)
+open class RxCollectionViewDataSourcePrefetchingProxy: DelegateProxy<UICollectionView, UICollectionViewDataSourcePrefetching>, DelegateProxyType, UICollectionViewDataSourcePrefetching {
```
Comment (kzaher) [inline RxCocoa/iOS/Proxies/RxCollectionViewDataSourcePrefetchingProxy.swift]: I think we use a bit different indent. Please take a look at `RxCollectionViewDataSourceProxy`.

File: RxCocoa/iOS/Proxies/RxTableViewDataSourcePrefetchingProxy.swift [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import UIKit
+import RxSwift
+
+@available(iOS 10.0, tvOS 10.0, *)
+extension UITableView: HasPrefetchDataSource {
+    public typealias PrefetchDataSource = UITableViewDataSourcePrefetching
+}
+
+@available(iOS 10.0, tvOS 10.0, *)
+let tableViewPrefetchDataSourceNotSet = TableViewPrefetchDataSourceNotSet()
```
Comment (kzaher) [inline RxCocoa/iOS/Proxies/RxTableViewDataSourcePrefetchingProxy.swift]: ... the same `fileprivate`.

File: RxCocoa/iOS/Proxies/RxTableViewDataSourcePrefetchingProxy.swift [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+@available(iOS 10.0, tvOS 10.0, *)
+final class TableViewPrefetchDataSourceNotSet : NSObject, UITableViewDataSourcePrefetching {
+
+    func tableView(_ tableView: UITableView, prefetchRowsAt indexPaths: [IndexPath]) {}
+
+}
+
+@available(iOS 10.0, tvOS 10.0, *)
+open class RxTableViewDataSourcePrefetchingProxy: DelegateProxy<UITableView, UITableViewDataSourcePrefetching>, DelegateProxyType, UITableViewDataSourcePrefetching {
```
Comment (kzaher) [inline RxCocoa/iOS/Proxies/RxTableViewDataSourcePrefetchingProxy.swift]: and indentation.

File: Tests/RxCocoaTests/UITableView+RxTests.swift [hunk 5]
```diff
@@ -218,6 +218,42 @@ final class UITableViewTests : RxTest {
...
+
+        let testIndexPaths = [IndexPath(item: 1, section: 0), IndexPath(item: 2, section: 0)]
+        tableView.prefetchDataSource!.tableView!(tableView, cancelPrefetchingForRowsAt: testIndexPaths)
+
+        XCTAssertEqual(indexPaths, testIndexPaths)
+        subscription.dispose()
+    }
+
     func test_delegateEventCompletesOnDealloc1() {
         let items: Observable<[Int]> = Observable.just([1, 2, 3])
```
Comment (kzaher) [inline Tests/RxCocoaTests/UITableView+RxTests.swift]: We are missing inheritance tests and dealloc tests. Please take a look at `test_DelegateEventCompletesOnDealloc` and `DelegateProxyTest+UIKit.swift`.

**Comment by kzaher** [comment]
I have no idea for your local test failures. Maybe you are using Xcode beta and failing because of #1555.

File: Tests/RxCocoaTests/UITableView+RxTests.swift (hunk 5 above)
Comment (RowbotNZ) [reply Tests/RxCocoaTests/UITableView+RxTests.swift]: Thanks for the feedback! I've had a go at copying the existing tests that you mentioned, please let me know if I've made a mistake.

**Pull request merged**